### DISCUSSION TITLE
关于自动养马的一些疑问

### DISCUSSION THREAD

**Discussion by fatinghenji** [NateScarlet/auto-derby#42] [Ideas]
### 部分问题是基于下面的文档
[ウマ娘育成理論 【汉化 by 天祈穹】.pptx](https://github.com/NateScarlet/auto-derby/files/6641563/by.pptx)
#### 看了上面的文章之后觉得定向养成确实很强！
结合自己目前使用脚本的体验，感觉定向性不是很强，即使用了3速3力的支援卡，也会在自动养成的时候把其他属性拉的很高。
- 是否可以在未来的版本中增添定向养成呢？优先拉取某X个属性到指定等级，如速A，耐A，然后再提升其他属性？或者提升X个属性的训练优先度？
- 在养成开始的时候选择自己的跑法偏好，如逃/追/先行
#### 在ura前进行技能升级 ~~已经在ura的总决赛上扑街了5、6次的人这样说到。~~
- 在ura前对技能进行升级可以在一定程度上提高ura的胜利机会，特别是在用了几次闹钟后还发现是2着的时候整个人都傻了:skull:
- 可以像对事件的选择一样，检测到ura比赛前提示技能升级，手动升级之后在terminal中直接敲回车继续脚本运行。
- 当然也可以直接敲回车然后跳过技能的升级。
#### 限时商店买空  ~~PCR后遗症让我已经无金币去升级支援卡~~
- 像题目中提到的一样，买空限时商店会导致金币啪就没了，导致没办法对支援卡进行升级。
- 在以后的版本中会增添是否买空的选项吗？
- 是否会支持仅购买马娘碎片/指定物品（如鞋子/闹钟）？
#### 对生涯中对粉丝要求有规定的马娘的养成效果很差
- ~~说的就是你ハルウララ~~
- 除了上面这个比较极端的例子，自动养成的过程中也会出现因为粉丝数目不够而出现脚本停止的情况。会在以后的版本中进行优化吗？
  - 直接跑比赛最上面的（双圈机率大）
  - 跑粉丝数多的，需要跑的总场数小
#### 跑法自动选择
- 除了出现**结果预见**~~好像是这个吧我日语真的好差~~的时候其他情况好像不能进行跑法的优选？
- 在跑法双A的情况下是怎么决定选择的呢？是根据人数？场地？还是什么？
- 在ura的时候（默认满闹钟，即可以使用4次），能否在出现双A跑法时在首选跑法跑3次未成功选择另一种跑法？
#### 事件选择
- 事件选择的结果是每个马娘不同还是只要是碰到相同事件就需要手动选择呢？
- 即使是相同的马娘选择的跑法/体力不同对事件的选择倾向也会不同。
#### 手机adb模式支持？
- 使用手机adb获取屏幕截图然后模拟点击，就可以把电脑空出来干别的了。
- 主要是现在的运行模式鼠标来回点击会造成打字/看视频的问题。

**Comment by Splendent** [comment]
手機自動點擊的話Robotmon不錯:P

**Comment by fatinghenji** [comment]
> 手機自動點擊的話Robotmon不錯:P

去搜索了下，每太弄明白怎么和现有的这个项目进行联动。

**Comment by NateScarlet** [comment]
这一个帖子发太多东西了 

我尽量把你说的提取成多个特性请求

新功能欢迎 PR
这个项目目前还只是针对我自己的使用情况的


## 定向养成

  如果要竞技场 我建议还是手动养，如果你重复一个套路多次需要自动化请描述方法然后我加新的工作（或者你参考nurturing.py 自己加）

## 在ura前进行技能升级

  #45 

## 限时商店买空

  #43 

## 对生涯中对粉丝要求有规定的马娘的养成效果很差

   #46 这个功能我在 README 中提了 但是还没实现

## 事件选择

  标题和选项都一样才会选择同样的选项，基本只有主线剧情事件会不同马娘共通。你可以每个马娘一个不同的 single_mode_choices.json， 启动器已经支持设置了。

## 手机adb模式支持

  #44 这个工作量比较大，我建议你直接用手机看视频

**Reply by fatinghenji** [reply]
我是个代码小白……只会用一点都不会写，所以PR真的无能为力……

**Comment by NateScarlet** [comment]
跑法和训练一样采用的是评分计算

https://github.com/NateScarlet/auto-derby/blob/409e0f307096308f9d4915a34f0810a52f22fa7a/auto_derby/jobs/nurturing.py#L182-L299

**Comment by fatinghenji** [comment]
@NateScarlet 关于培养这部分的计算，可以试一下下面这个方法。
- 我们用P(Point)来作为标准单位。
- 不同于目前的计算方法，我们需要新增对于5大数值的倾向性参数：最优先，次优先，其他。
- 根据氪金量的不同分为微氪（零氪）、中度氪金、重度氪金。（注：每SSR每1破记作新抽到1SSR卡）
  - 微氪（零氪）：SR卡大队，SSR卡小于15张
  - 中度氪金：SSR卡16-45张
  - 重度氪金：SSR卡45张以上
- 下面的一些数据来源参考的是一楼给出的PPT。
- 每次计算所需要考虑的参数：能力值的增加、羁绊值的增加、体力的减少、技能点的增加、凭借提示、心情的变化、机会发生差分
  - 能力值的改变：
    - 每1点为1Point
  - 羁绊值的增加——因为触发彩圈需求的羁绊值最小为4格，所以需要分开计算：
    - 当羁绊值不满4格时：
      - 主要训练的能力：每有一个助训者记作8P
      - 除主要训练的能力：每有一个助训者记作5P
    - 当羁绊值达到四格时，记作0P
  - 体力的改变：
    - 微氪（零氪）：5体力=3P
    - 中度氪金：3体力=2P
    - 重度氪金：5体力=4P
  - 技能点的增加：5pt=2P
  - 特殊事件（出现❗)：每一个❗记作5P
  - 心情变化（用来评判要不要出去玩）——30体力=1心情变化：
    - 零氪（微氪）：18P
    - 中度氪金：20P
    - 重度氪金：24P
  - 机会发生差分（~~我觉得这个名字好怪~~）：
    - 最优先训练属性：是40P
    - 次优先训练属性：是30P
    - 其他训练属性：15P
    - 重氪在上面的值增加10P，零氪（微氪）减少5P。
- 计算方法
    - 将上述各参数相加
    - 有训练概率时（例子可以参考PPT里面的第26张）
      - 成功为+，失败为-，需要按几率计算加权平均数。
      - 主要对应低体力的时候是梭哈一次还是老老实实休息。

**Comment by fatinghenji** [comment]
上面这个算法的问题是怎么读取有几个人？

**Comment by NateScarlet** [comment]
现在的你能养出五边形战士？ 不应该是速力吗

**Reply by fatinghenji** [reply]
一个非常不恰当的例子……我的问题

**Reply by fatinghenji** [reply]
删掉了

**Comment by NateScarlet** [comment]
训练算法我在 #14 说过了 
也是评分制的 不过和你说的这些细节不一样

**Reply by fatinghenji** [reply]
找了个孤例，比如我在[#20](https://github.com/NateScarlet/auto-derby/issues/20)里面的截图，套用上面的公式结果如下：
训练内容 | 速度 | 耐力 | 力量 | 毅力 | 智力 | 技能点 | 不满4格数量 | 体力 | 感叹号 | 心情 | 机会发生差分 | 总分
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
速度 | 15 | 0 | 6 | 0 | 0 | 2 | 1 | -14 | 0 | 0 | 40 | 61.4
耐力 | 0 | 18 | 0 | 6 | 0 | 3 | 0 | -14 | 0 | 0 | 15 | 31.8
力量 | 0 | 8 | 11 | 0 | 0 | 2 | 0 | -14 | 0 | 0 | 30 | 41.4
毅力 | 5 | 0 | 5 | 13 | 0 | 2 | 0 | -14 | 0 | 0 | 15 | 30.4
智力 | 3 | 0 | 0 | 0 | 15 | 3 | 1 | 10 | 0 | 0 | 15 | 45.2

按照[#14](https://github.com/NateScarlet/auto-derby/discussions/14)里面的算法，我把速度设定为第一优先，力量设定为第二优先。
结果显示应该进行速度训练。当然这是非常不可靠的孤例，明天我试试用相同的配卡和种马用这两种算法分别养一只。

**Reply by NateScarlet** [reply]
这算法我觉得不靠谱 因为不考虑当前属性影响 赛马娘里面属性增加的效果不是线性的而是边际递减的

**Reply by NateScarlet** [reply]
尤其是耐力 如果你其他的训练一直超好导致耐力一点不加那育成中途就失败了

**Reply by fatinghenji** [reply]
> 尤其是耐力 如果你其他的训练一直超好导致耐力一点不加那育成中途就失败了

是这样，刚刚跑了下真就没点一次耐力。

**Comment by NateScarlet** [comment]
自动育成的目的不是竞技场 竞技场只能手动练

刷因子就是得水桶 600+属性才可能出现三星因子

建议你能够全程不点技能URA胜利再用自动育成

**Comment by NateScarlet** [comment]
你靠人数判断意义不大 有的支援卡4羁绊是为了金技能 有的就只是为了训练加成

金技能需要具体到某张卡，人数没用 训练加成的人数判断我一开始准备做然后发现用训练效果就足够了 
如果要识别只需要用最后一格羁绊的图模版匹配就行了

还有一个增强的是根据训练级别，需求训练没到5级在前期有额外加分 这个可以做但是优先级比较低

**Reply by fatinghenji** [reply]
技能这个在PPT里面有讲：
`根据对技能的需求程度在3~10p之间调整　也可以统一记作5p(简单计算)
`
计算人数：
1、是为了更早的达到触发“友情タッグ”条件，详细说明可以看PPT的P15（详细计算版），简单来说就是：**在确保倾向性加成的情况下，用更少的训练次数来达到“友情タッグ”**
2、你说的情况确实是这样，因为人数加成是直接反应在训练数值里面的。

**Reply by NateScarlet** [reply]
用当前获取较少属性来换取之后有几率获取更高属性是赚是亏很难说 更何况不根据具体卡做判断统一用一个评分效果就更差了
不如直接多育成几次赌它直接在你要的属性上加羁绊

**Comment by fatinghenji** [comment]
现在预判比赛胜利的方法是什么呢？
- 看了下log文件，发现依旧是评分体系，但是不知道这个评分是怎么得出的。
```
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<函館 turf 1800m OP 巴賞>, order=11, difficulty=832, best_style_score=252.45
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<函館 dart 1700m OP マリーンステークス>, order=7, difficulty=1102, best_style_score=208.60
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<中京 dart 1800m OP 名鉄杯>, order=7, difficulty=1102, best_style_score=229.47
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<中京 turf 1200m G3 CBC賞>, order=8, difficulty=945, best_style_score=211.71
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<中京 dart 1400m G3 プロキオンステークス>, order=5, difficulty=1215, best_style_score=174.80
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<福島 turf 2000m G3 七夕賞>, order=1, difficulty=675, best_style_score=601.39
DEBUG [2021-06-16 22:28:16]:auto_derby.single_mode.race:319: estimate order: race=Race<函館 turf 2000m G3 函館記念>, order=1, difficulty=675, best_style_score=656.06
```
- 今天逛NGA论坛的时候发现了个比赛模拟器，感觉模拟的结果更准确一点（可以包括技能影响）。
  - 原帖地址：[[游戏攻略] 赛马模拟器alpha版释出](https://bbs.nga.cn/read.php?tid=27099540)
  - 前端地址：[ウマ娘レースエミュレーター](http://race.wf-calc.net/)
  - 开源地址：[urakagi/uma-clock-emu](https://github.com/urakagi/uma-clock-emu)）
- 能适配进目前的程序吗？
  - 支持的理由：
    - 截屏可以读取出目前拥有的技能。
    - 如果升级完技能后依旧无法跑第一，那么就去休息（或者外出旅行）。
  - 不支持的理由：
    - 【重要】上面是以时间为标准来进行结果输出的，怎么判断当前比赛里面其它马娘跑的时间呢？还有以下几个小问题
      - 育成里面比赛的npc马娘是相同的吗？
      - 相同的比赛会有不同的场地天气影响吗？(好像不用考虑这个，因为可以通过ocr读到)
    - 没有在代码页看到开源协议
    - 语言是Vue，本项目是python

**Reply by NateScarlet** [reply]
auto_derby.single_mode.race:319 代表 auto_derby/single_mode/race.py 第 319 行

你发的这个是手动选技能 没看到截屏读取

赛事预测还在调 我写出来一共也没跑几局 见 https://github.com/NateScarlet/auto-derby/discussions/61#discussioncomment-879091

**Reply by fatinghenji** [reply]
> 你发的这个是手动选技能 没看到截屏读取

是我表述的问题，我的意思是通过目前auto-derby的ocr功能，识别出已经拥有的技能/数值/比赛，然后填充入我上面提到的程序进行计算，计算结果返回到auto-derby里面，选择是否要参加比赛。

**Reply by NateScarlet** [reply]
自动化浏览器操作工作量比直接用 python 重写它的算法还多

它的模拟算法应该在 [Main.vue](https://github.com/urakagi/uma-clock-emu/blob/5ac72d143c4d45157860db68bb1fd8ea4f47cadb/src/components/Main.vue) 和UI逻辑混在一起一个文件 1500 行……

问题是我不知道它算法和数据怎么来的 

用统计应该能实现差不多的效果 #62

**Reply by fatinghenji** [reply]
数据和算法来源可能是解包，在原贴中作者提到过自己不会解包，但是后来有人pr了解包的相关数据。

**Reply by fatinghenji** [reply]
>问题是我不知道它算法和数据怎么来的

nga论坛上的另一个帖子里面有说明参数的使用方法，但是不清楚是否和上面的项目使用相同的计算方法。
[我的马娘为什么输了？属性和技能对比赛具体有哪些影响](https://bbs.nga.cn/read.php?tid=26010713)

**Reply by NateScarlet** [reply]
https://github.com/NateScarlet/auto-derby/blob/96112ab57fe6c22e6c029d90d4aa6a72da7f1fd6/auto_derby/single_mode/race.py#L74

**Comment by fatinghenji** [comment]
#62的想法不错，但是有几个问题无法解决。
- 马娘的技能是否触发？
  - 无法通过屏幕截图来判断（似乎有文字版的赛程记录？我记不太清楚）
  - 可以通过抓取通讯包读取马娘跑完比赛的结果（包括体力，速度。技能情况），但是不清楚是否违反cy的用户协议。
- 相同的比赛使用不同的马娘跑所带技能不同，怎么量化？
- 记录在服务器成本怎么解决？记录在本地收集数据时间是否太长？（每天跑3次比赛）
  - 服务器考虑捐赠？
  - 本地数据考虑通过github的pr和release进行数据更新和增添？(需要有一定的代码基础？)

**Reply by NateScarlet** [reply]
无视技能 因为脚本也不会自动点技能

每天不是 3 次比赛而是 3 * 20 次比赛

我自己有服务器 很多数据库服务也有免费额度 不过数据量少的情况下先没必要用服务器 遇到离谱的预测提 issue 时顺便带上就行了

**Comment by NateScarlet** [comment]
商店默认行为改成了请求人工处理 
自动买所有物品需要使用插件 limited_sale_buy_everything f1e98fdbbcde678934e83269176397e754ee338c
